### DISCUSSION TITLE
Pods with imagemagick are restarting in Kubernetes after certain intervals

### DISCUSSION THREAD

**Discussion by mangudaigb** [ImageMagick/ImageMagick#3301] [Help]
Hi everyone
I have a java web application which makes calls to ImageMagick to resize images dynamically during response time. The application uses IM4Java to make the _convert_ call to ImageMagick. The problem is the RSS memory increases over time and when the memory crosses the pod limit, the OOM Killer becomes active and kills the pod. 
The pod limits are: 6 GB and 2048 m
<img width="923" alt="image" src="https://user-images.githubusercontent.com/5712772/109206342-bd212880-77cd-11eb-8626-d6d1f10acba0.png">

Analysis of the Java heap shows that java usage is around ~1.6 GB. The java options for heap include 
--Xms1024m --Xmx2048m -XX:MaxMetaspaceSize=256m -XX:NewRatio=1 -XX:+UseConcMarkSweepGC. From the java process analysis, the total memory usage for java was about ~1.6 GB.

The RSS memory during this time had increased to 4.3 GBs. 

Most of the operations on ImageMagick are just resize commands.

Any pointers on fixing is highly appreciated? 

Regards
Jibi

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/architecture.php#tera-pixel.  Even though you likely are not using large images, the principles are the same.  Set the memory and map limits to 256MiB, for example.  Any image that requires more memory than your limits are pushed to disk.  Its slower but it caps the amount of memory ImageMagick consumes.

**Reply by mangudaigb** [reply]
I did set the memory and map limits to 256 MiB but to no avail.

**Comment by urban-warrior** [comment]
We can only address an issue if its traced directly to ImageMagick.  Post a sample command-line here along with links to any image you utilize in the command-line.  We will download the images and run your command-line while measuring exactly how much memory is consumed.   If the memory requirements are excessive, we have possible methods to reduce the memory requirements.  We will also confirm there are no memory leaks from from your command-line.

**Reply by mangudaigb** [reply]
Image: https://i.postimg.cc/vH5tf0mf/image-1617089551.jpg
Command: convert -resize 540x540! +repage <source_image> <destination_image>

And, thanks a lot for your help.

**Reply by urban-warrior** [reply]
Try this command:

```
convert -limit memory 512MiB -limit map 1GiB -define jpeg:size=540x540 image-1617089551.jpg -resize 540x540! +repage image-1617089551.png
```

If the command still exceeds memory, divide the memory and map limits in half (e.g. 256MiB).  If that still fails, try increasing your Java heap memory.

**Reply by mangudaigb** [reply]
Hi Cristy
Sorry for the delayed response, I was doing some tests on my end as the memory options did not quite pan out. And after a lot of head banging, I found that it was because of how the java framework interacts with the Java system. Sorry for the trouble and thanks for all the help. 
Cheers
